tools/c7n_azure/tests/test_storage.py

```diff
@@ -42,3 +60,172 @@ def test_value_filter(self):
+
+        vnets = p_vnet_get.run()
+
+        id1 = vnets[0]['properties']['subnets'][0]['id']
+        id2 = vnets[1]['properties']['subnets'][0]['id']
+
+        p_add = self.load_policy({
+            'name': 'test-azure-storage-add-ips',
+            'resource': 'azure.storage',
+            'filters': [
+                {'type': 'value',
+                 'key': 'name',
+                 'op': 'glob',
+                 'value_type': 'normalize',
+                 'value': 'cctstorage*'}],
+            'actions': [
+                {'type': 'set-network-rules',
+                 'default-action': 'Deny',
+                 'bypass': ['Logging', 'Metrics'],
+                 'virtual-network-rules': [
+                     {'virtual-network-resource-id': id1},
+                     {'virtual-network-resource-id': id2}
+                 ]}
+            ]
+        })
+
+        p_add.run()
+
+        resources = self._get_resources()
+        self.assertEqual(len(resources), 1)
+        rules = resources[0]['properties']['networkAcls']['virtualNetworkRules']
+        self.assertEqual(len(rules), 2)
+        self.assertEqual(rules[0]['id'], id1)
+        self.assertEqual(rules[1]['id'], id2)
+        self.assertEqual(rules[0]['action'], 'Allow')
+        self.assertEqual(rules[1]['action'], 'Allow')
+
+    @arm_template('storage.json')
+    def test_empty_bypass_network_rules_action(self):
+        p_add = self.load_policy({
+            'name': 'test-azure-storage-add-ips',
+            'resource': 'azure.storage',
+            'filters': [
+                {'type': 'value',
+                 'key': 'name',
+                 'op': 'glob',
+                 'value_type': 'normalize',
+                 'value': 'cctstorage*'}],
+            'actions': [
+                {'type': 'set-network-rules',
+                 'default-action': 'Deny',
+                 'bypass': []}
+            ]
+        })
+
+        p_add.run()
+
+        resources = self._get_resources()
+        bypass = resources[0]['properties']['networkAcls']['bypass']
+        self.assertEqual(bypass, 'None')
+
+    @arm_template('storage.json')
+    def test_missing_bypass_network_rules_action(self):
+        p_add = self.load_policy({
+            'name': 'test-azure-storage-add-ips',
+            'resource': 'azure.storage',
+            'filters': [
+                {'type': 'value',
+                 'key': 'name',
+                 'op': 'glob',
+                 'value_type': 'normalize',
+                 'value': 'cctstorage*'}],
+            'actions': [
+                {'type': 'set-network-rules',
+                 'default-action': 'Deny'}
+            ]
+        })
+
+        p_add.run()
+
+        resources = self._get_resources()
+        bypass = resources[0]['properties']['networkAcls']['bypass']
+        self.assertEqual(bypass, 'None')
+
+    @arm_template('storage.json')
+    def test_bypass_network_rules_action(self):
+        p_add = self.load_policy({
```

Comment:
this check looks like it could be collapsed into one of the previous tests re `test_network_ip_rules_action` to just verify the bypass setting there.

Reply:
A test must have only one reason why it can fail. It is subjective, but in my opinion it is always better to have a separate test for a separate feature. Here bypass has a different logic than ip rules. The logic for handling bypass in the policy might change independently from ip rules and if something is broken it is the bypass test that must fail not ip_rules test.